### PR TITLE
Upgrade wagmi, viem, web3modal

### PR DESCRIPTION
**Pull request by mudrila**
This PR migrates from `wagmi` v1 to `wagmi` v2 and upgrades related packages version, including following:

- `viem` (upgraded to `v2`)
- `@web3modal/wagmi` (upgraded to `v4`)
- `@tanstack/react-query` (added since it is `peerDependency` of `wagmi`)
- `Node.js` (upgraded to current LTS, `20.11.1`)
- `@walletconnect/ethereum-provider` (removed)

This PR does not include removing `ethers` nor upgrading that - comes in scope of further implementation.
Also solves https://github.com/decentdao/fractal-interface/pull/1465#pullrequestreview-1942944244

### CODE REVIEW

File: src/components/ui/menus/AccountDisplay/MenuItemNetwork.tsx [hunk 1]
```diff
@@ -43,7 +45,7 @@ export function MenuItemNetwork() {
           cursor="pointer"
           onChange={async e => {
             e.preventDefault();
-            switchNetwork(Number(e.target.value));
+            switchChain({ chainId: Number(e.target.value) });
```
Comment (adamgall) [inline src/components/ui/menus/AccountDisplay/MenuItemNetwork.tsx:48]: Switching to Polygon causes the app to crash.

**Review by adamgall — CHANGES_REQUESTED** [review]

**Comment by adamgall** [comment]
@mudrila in these commits, there are an awful lot of changes to the `package-lock.json` file. This makes sense due to various packages being upgraded. However I'm curious: were there any other reasons for the changes to this lockfile? Things including, but not limited to:
- running `npm update`
- deleting the `package-lock.json` file and letting it regenerate via `npm install`
- anything else?

File: src/hooks/utils/useUpdateTimer.tsx [hunk 2]
```diff
@@ -3,7 +3,7 @@ import { useIdleTimer } from 'react-idle-timer';
 
 export const useUpdateTimer = (safeAddress?: string | null) => {
   // Use useRef to store timers without triggering unnecessary re-renders
-  const timers = useRef<Map<string, NodeJS.Timer>>(new Map());
+  const timers = useRef<Map<string, NodeJS.Timeout>>(new Map());
```
Comment (adamgall) [inline src/hooks/utils/useUpdateTimer.tsx:6]: As far as I can tell, the project builds and runs fine when this is kept as `NodeJS.Timer`

File: src/hooks/utils/useMasterCopy.ts [hunk 3]
```diff
@@ -53,16 +53,16 @@ export function useMasterCopy() {
...
       const filter = contract.filters.ModuleProxyCreation(proxyAddress, null);
       return contract.queryFilter(filter).then(proxiesCreated => {
         if (proxiesCreated.length === 0) {
-          return [constants.AddressZero, 'No proxies created'] as const;
+          return [constants.AddressZero, 'No proxies created'];
         }
         const masterCopyAddress = proxiesCreated[0].args!.masterCopy;
         setValue(CacheKeys.MASTER_COPY_PREFIX + proxyAddress, masterCopyAddress);
-        return [masterCopyAddress, null] as const;
+        return [masterCopyAddress, null];
```
Comment (adamgall) [inline src/hooks/utils/useMasterCopy.ts:65]: As far as I can tell, the project builds and runs just fine without these changes.

File: src/hooks/utils/useAvatar.ts [hunk 4]
```diff
@@ -1,10 +1,12 @@
...
+import { useEnsAvatar } from 'wagmi';
+import { useNetworkConfig } from '../../providers/NetworkConfig/NetworkConfigProvider';
 
 const useAvatar = (account: string | null) => {
-  const { chain } = usePublicClient();
+  const { chainId } = useNetworkConfig();
   const { data: avatarURL } = useEnsAvatar({
-    name: account,
-    chainId: chain.id,
+    name: account as Address,
```
Comment (adamgall) [inline src/hooks/utils/useAvatar.ts:8]: This `account` should be typed as an `Address` in the hook's input parameters, and that change is propogated back through all uses of this hook in the app.

File: src/components/ui/proposal/useProposalCountdown.tsx [hunk 5]
```diff
@@ -38,7 +38,7 @@ export function useProposalCountdown(proposal: FractalProposal) {
     governanceDispatch: action.dispatch,
   });
 
-  let updateStateInterval = useRef<NodeJS.Timer | undefined>();
+  let updateStateInterval = useRef<NodeJS.Timeout | undefined>();
```
Comment (adamgall) [inline src/components/ui/proposal/useProposalCountdown.tsx:41]: As far as I can tell, the project builds and runs just fine without these changes.

File: src/components/ui/proposal/useProposalCountdown.tsx [hunk 6]
```diff
@@ -84,7 +84,7 @@ export function useProposalCountdown(proposal: FractalProposal) {
   }, [secondsLeft, loadDAOProposals, proposal, updateProposalState, governance.type, dao]);
 
   useEffect(() => {
-    let countdownInterval: NodeJS.Timer | undefined;
+    let countdownInterval: NodeJS.Timeout | undefined;
```
Comment (adamgall) [inline src/components/ui/proposal/useProposalCountdown.tsx:87]: As far as I can tell, the project builds and runs just fine without these changes.

File: src/components/ui/proposal/useProposalCountdown.tsx [hunk 7]
```diff
@@ -161,7 +161,9 @@ export function useProposalCountdown(proposal: FractalProposal) {
     getCountdown();
 
     return () => {
-      clearInterval(countdownInterval);
+      if (countdownInterval) {
+        clearInterval(countdownInterval);
+      }
```
Comment (adamgall) [inline src/components/ui/proposal/useProposalCountdown.tsx:166]: As far as I can tell, the project builds and runs just fine without these changes. `clearInterval` accepts `undefined` so this should be fine without the conditional check.

File: src/components/ui/forms/BigNumberInput.tsx [hunk 8]
```diff
@@ -19,6 +19,7 @@ export interface BigNumberInputProps
   min?: string;
   max?: string;
   maxValue?: BigNumber;
+  placeholder?: string;
```
Comment (adamgall) [inline src/components/ui/forms/BigNumberInput.tsx:22]: Why was this added?

File: src/components/ui/forms/ABISelector.tsx [hunk 9]
```diff
@@ -34,7 +34,7 @@ export default function ABISelector({ target, onChange, onFetchABI }: IABISelect
 
   useEffect(() => {
     const loadABI = async () => {
-      if (target && ((ensAddress && isAddress(ensAddress)) || isAddress(target))) {
+      if (client && target && ((ensAddress && isAddress(ensAddress)) || isAddress(target))) {
```
Comment (adamgall) [inline src/components/ui/forms/ABISelector.tsx:37]: Why was `client` added to this list?

**Review by adamgall — COMMENTED** [review]
A handful of comments. In addition to the line-by-line comments, a big other question I have is:

In these commits, there are an awful lot of changes to the `package-lock.json` file. This makes sense due to various packages being upgraded. However I'm curious: were there any other reasons for the changes to this lockfile? Things including, but not limited to:
- running `npm update`
- deleting the `package-lock.json` file and letting it regenerate via `npm install`
- anything else?

File: tsconfig.json [hunk 10]
```diff
@@ -17,6 +17,6 @@
     "jsx": "preserve",
     "incremental": true
   },
-  "exclude": ["docker"],
-  "include": ["src", "tests", "test", "app", "netlify/functions", "vite.config.mts"]
+  "exclude": ["docker", "netlify/functions"],
+  "include": ["src", "tests", "test", "app", "vite.config.mts"]
```
[the remark this answers is not in the document]
Comment (adamgall) [reply tsconfig.json]: I'm able to build with "netlify/functions" in the "include" section here. Are you not?

File: src/providers/Ethers/index.tsx [hunk 11]
```diff
@@ -22,20 +22,22 @@ export default function EthersContextProvider({ children }: { children: ReactNod
   const { data: walletClient } = useWalletClient();
 
   const provider = useMemo(() => {
```
[the remark this answers is not in the document]
Comment (adamgall) [reply src/providers/Ethers/index.tsx:24]: Ah yes, I see now! Thanks.

File: src/hooks/utils/useAvatar.ts (hunk 4 above)
Comment (adamgall) [reply src/hooks/utils/useAvatar.ts]: Got it, thanks.

**Comment by adamgall** [comment]
> > @mudrila A handful of comments. In addition to the line-by-line comments, a big other question I have is:
> > In these commits, there are an awful lot of changes to the package-lock.json file. This makes sense due to various packages being upgraded. However I'm curious: were there any other reasons for the changes to this lockfile? Things including, but not limited to:
> 
> * running `npm update`
> * deleting the `package-lock.json` file and letting it regenerate via `npm install`
> * anything else?
> 
> So those massive changes were caused by me deleting the `package-lock.json` and re-generating that. That was the reason we were getting newest `TypeScript` version and build failures caused by that. That's also was leading to CI build to fail. When I realized this - I copied over `package-lock.json` from `develop`, executed `npm install` to get updates from newest versions of packages (`wagmi`, `viem`, `web3modal` etc.) and pushed updated lock file. That solved the problem with these redundant changes and CI build failure.
> 
> TBH I wasn't expecting anything "bad" from deleting `package-lock.json` - I remember that was the case for `CRA` but with `Next.js` that wasn't happening and my idea was that with `Vite` there won't be any problems - but yeah, it broke the build, lessons learned 😄

Ok, makes sense. In general, it's a bad idea to delete `package-lock.json` and regenerate it from scratch. That will likely cause lots of changes to package versions that are unintended, and can wreak havoc on a large codebase with many dependencies due to so many packages changing at once.

Glad to hear that was reverted.

**So @mudrila to be clear, the only changes to `package-lock.json` are the result of explicitly updating the packages we wanted to update, right?**

File: package-lock.json
Comment (adamgall) [inline package-lock.json:1]: While we're here, might as well remove the `version` property from this file, and update it to `lockfileVersion: 3` (can be done by running `npm install --lockfile-version 3`).

**Review by adamgall — CHANGES_REQUESTED** [review]

**Review by Da-Colon — APPROVED** [review]

**Pull request merged**